### ISSUE TITLE
replaceData() scrolls to top

### ISSUE THREAD

**Issue by Albot15** [olifolkerd/tabulator#4155]
Hello Tabulator community,

I am relatively new to Tabulator and have been experiencing an issue with the scrolling feature. I have a table that updates its data via AJAX every minute using the replaceData() function. However, whenever the AJAX triggers, the table resets its scroll position to the top, despite the documentation stating that it should preserve the scroll position and other filters.

I would appreciate it if someone could help me with this issue, as I need to retain the scroll position after the table updates. 

I am using 5.4.4

Code:

` var table = new Tabulator("#tabutable", {
        pagination: "local",
        paginationSize: 100,
        initialSort: [{
            column: "index",
            dir: "asc"
        }],
        columns: [ //Define Table Columns
            {
                title: "#",
                field: "index",
                formatter: "rownum",
                align: "center",
                width: 50
            },
...
        ],
        ajaxURL: "ajaxurl",
        layout: "fitColumns",
        responsiveLayout: "collapse",

    });`

Thank you in advance for your help.
Best regards,

**Comment by lekoala** [comment]
yes, that's because when you replace the data, you empty all the rows (table has height 0, and page scroll), then insert it again.

i had the exact same issue. my fix was to add a min height to the table :-)

**Comment by Albot15** [comment]
Hi,

Thanks for your help @lekoala . I just tested with minHeight and it works. However, its not a perfect solution. If the total number of rows is less than the height of the table, it will leave a big empty space. For now, your solution is great.

If anyone knows how we can update the table via Ajax without affecting the scroll position, please let me know.

Thanks again!

**Comment by lekoala** [comment]
yes, actually it would be pretty simple to add the logic into the core, i did something similar for the datatree plugin where it goes like this. Leaving this for reference, obviously that's not the "proper" way of doing things

            // prevent redraw due to size change
            row.table.element.style.minHeight =
                row.table.element.offsetHeight + "px";

            row.reinitialize();

            this.refreshData(true);

            this.dispatchExternal(
                "dataTreeRowCollapsed",
                row.getComponent(),
                row.modules.dataTree.index
            );

            // adjust to collapsed size
            row.table.element.style.minHeight =
                row.table.element.querySelector(".tabulator-tableholder")
                    .offsetHeight + "px";

**Comment by arpsmack** [comment]
Fiddle that demonstrates the issue and has some super hacky workarounds: https://jsfiddle.net/c4hwu7qf/1/

**Comment by lekoala** [comment]
So i've made a better solution that doesn't require hacking


        // Fix table size on full redraw
        tabulator.on("renderStarted", () => {
            tabulator.element.style.minHeight =
            tabulator.element.offsetHeight + "px";
        });
        tabulator.on("renderCompleted", () => {
            tabulator.element.style.minHeight =
            tabulator.element.querySelector(".tabulator-tableholder")
                    .offsetHeight + "px";
        });

**Comment by arpsmack** [comment]
That looks like a pretty clean solution, but I tried it in the fiddle and it causes an infinite loop. I think you need to use the `clientHeight` of the table element in the `renderStarted`, otherwise this check gets hit https://github.com/olifolkerd/tabulator/blob/618e303fe50fdb8c6a7683a436c598ad7118447e/src/js/core/RowManager.js#L1055-L1062 causing an immediate redraw leading to the infinite loop.

Although I guess this depends on your CSS. In my case, `offsetHeight` is 2px larger than `clientHeight` because there's a 1px border.

Also, even if you fix that issue, it still has the jankyness of jerking a bit every time `replaceData` happens in Firefox. Seems to work pretty well in Chrome unless you scroll all the way down to where the pagination controls are visible, in which case every time `replaceData` happens it jerks the pagination controls back out of view.

**Comment by lekoala** [comment]
It's tricky... for firefox i've discovered that `overflowAnchor` is what's causing the issue

my updated snippet

    tabulator.on("renderStarted", () => {
      const holder = tabulator.element.querySelector(".tabulator-tableholder");
      holder.style.minHeight = holder.clientHeight + "px";
      holder.style.overflowAnchor = "none";
    });
    tabulator.on("renderComplete", () => {
      const holder = tabulator.element.querySelector(".tabulator-tableholder");
      const table = tabulator.element.querySelector(".tabulator-table");
      // If you want to keep table with paginated element with the same size if rows are missing (eg: on last page)
      let paginatedHeight = 0;
      if (this.config.paginationSize) {
        paginatedHeight = this.config.paginationSize * table.firstChild.offsetHeight;
      }
      holder.style.minHeight = Math.max(Math.min(holder.clientHeight, table.clientHeight), paginatedHeight) + "px";
    });

**Comment by phoenixstudiodz** [comment]
I checked the scroll position on `renderStarted` and `renderComplete` and the window.scrollY was the same.
I think what causing this behavior in `react-tabulator` is the table height, during update and for a brief moment all data in table is gone and it reduce the size.
Page size reduces => no need for scroll => browser scroll value will change.
Two possible temporary fix can be used (Both worked for me)

1- set `CSS` `minHeight` to the value `clientHeight` of tabulator => during this update, when table is empty it will retain it's size and there will be no scroll.
2- store current scroll value on `rowUpdated` and call `window.scrollTo` on `renderComplete` event (scroll `behavior` need to be `auto` or 'instant`).

The first solution should be better however the table height will not "shrink" after deleting rows.

**Comment by hayzamjs** [comment]
I ran into the same issue—tables that completely replace their data always lose scroll position, which is really frustrating. To work around it, I borrowed @arpsmack’s hack: before you replace the data, grab and save the container’s scrollTop and scrollLeft; after the update (renderComplete), reset them to the saved values.

If it helps someone, [here](https://github.com/AlchemillaHQ/Sylve/blob/master/web/src/lib/components/custom/TreeTable.svelte#L142) is my implementation. It's in svelte but the tabulator logic should be the same regardless.
